papers/bradley_dice/55_bradley_dice.rst

```diff
@@ -0,0 +1,584 @@
+
+..
+    TODO: Categorize papers by field, show counts? e.g. The most common scientific fields citing signac are materials science (10), molecular simulation (8), optical materials (5), ...
+
+    Brandon - this is cool, but may be more work than is worth it.
+
+While there is no "typical" **signac** project, factors such as computational complexity and data sizes offer some rough guidelines for when **signac**'s database-on-the-filesystem is appropriate.
+For instance, the time to check the status of a workflow depends on the number of jobs, number of operations, and number of conditions to evaluate for those jobs.
+To give a rough idea of the limits of scalability, it can be difficult to scale **signac** projects beyond around 100,000 jobs while keeping tasks like checking workflow status in an "interactive" time scale of 1-2 minutes.
+Many **signac** projects have 100 to 10,000 jobs, with each job workspace containing arbitrarily large data sizes (the total file size of the job workspace has little effect on the speed of the **signac** framework).
+Some users that primarily wish to leverage **signac-flow**'s workflows for execution and submission may have a very small number of jobs (< 10).
+One example of this would be executing a small number of expensive biomolecular simulations using different random seeds in each job's state point.
+Importantly, projects with a small number of jobs can be expanded at a later time, and make use of the same workflow defined for the initial set of jobs.
+The abilities to grow a project and change its schema on-the-fly catalyze the kind of exploration that is crucial to answering research questions.
+
+The workflow submission features of **signac-flow** interoperates with popular HPC schedulers including SLURM, PBS/TORQUE, and LSF automating the generation and submission of scheduler batch scripts.
+Directives are set through Python function decorators and define resource and execution requests for operations.
+Examples of directives include number of CPUs or GPUs, the walltime, and memory.
+The use of directives allows **signac-flow** workflows to be portable across HPC systems by generating resource requests that are specific to each machine's scheduler.
+
+Overview of New Features
+------------------------
+
+The last three years of development of the **signac** framework have expanded its usability, feature set, user and developer documentation, and potential applications.
+Some of the largest architectural changes in the framework will be discussed in their own sections, namely extensions of the workflow model (support for executing groups of operations and aggregators that allow operations to act on multiple jobs) and a much more performant and flexible re-implementation of the core "data structure" classes that synchronize **signac**'s Python representation of state points and job documents with JSON-encoded dictionaries on disk.
```

Comment:
Sort of. The optimizations that I did for signac 1.5 / 1.6 were also incorporated into the design of synced collections. There was some amount of optimization work happening concurrently in the existing code and synced collections, so the performance of synced collections is similar to the optimized "original" code in signac 1.6 but both are significantly faster than signac <= 1.4.

Reply:
Ok, the re-implementation then includes both the synced-collections and some other refactorizations.

Reply:
@bdice @csadorf just so you're both fully aware, here's a more thorough overview of the performance implications of various changes:
- Many lazy-loading-related changes happened concurrently with synced collections, so those performance numbers are mixed in a bit.
- All core operations on the new `JSONDict` based on synced collections are 1.5-3x faster than operations on the old `JSONDict` depending on the operation (mostly from #453, but also a few other PRs).
- `JSONDict` creation is about 1.5x slower with the new class simply because it has many more attributes created on construction, e.g. for thread-safety and because we no longer use `contextlib.contextmanager` but instead create contexts on construction (which pays off quickly if we operate on the object). 
- The new buffered mode I implemented in #462 is dramatically faster than the old buffered mode on fast file systems (1-2 orders of magnitude), but is not yet enabled on signac master. I plan to make a PR to change that soon.

The upshot is that for the purpose of signac-flow the synced collections-based `JSONDict` is marginally slower than the old `JSONDict` at present (without the new buffering enabled) because creating objects is the most common operation and we rarely modify them, just loop over all jobs. For user workflows it will depend a lot on how people use the objects, but in most cases synced collections should be faster. However, as Bradley says all the other optimizations that happened made it so that the end result was still faster than previous releases in all cases.

Note: one way in which we can substantially speed things up further would be to remove attribute-based access to state point/document elements, but that's a non-starter in the short to intermediate term.